src/types/edge-spec.ts

```diff
@@ -1,18 +1,67 @@
-import type { EdgeSpecRouteFn, EdgeSpecRouteParams } from "./web-handler"
+import {
+  createEdgeSpecRequest,
+  type EdgeSpecRouteFn,
+  type EdgeSpecRouteParams,
+} from "./web-handler.js"
 
-export type EdgeSpecRouteMatcher = (pathname: string) => {
-  matchedRoute: string
-  routeParams: EdgeSpecRouteParams
-}
+import type { ReadonlyDeep } from "type-fest"
+
+export type EdgeSpecRouteMatcher = (pathname: string) =>
+  | {
+      matchedRoute: string
+      routeParams: EdgeSpecRouteParams
+    }
+  | undefined
+  | null
 
-export interface EdgeSpec {
+export type EdgeSpecRouteMap = Record<string, EdgeSpecRouteFn>
+
+export interface EdgeSpecOptions {
   routeMatcher: EdgeSpecRouteMatcher
-  routeMapWithHandlers: {
-    [route: string]: EdgeSpecRouteFn
-  }
+  routeMapWithHandlers: EdgeSpecRouteMap
+
+  handleModuleServiceRouteNotFound?: EdgeSpecRouteFn
+  handle404?: EdgeSpecRouteFn
 }
 
-export type EdgeSpecAdapter<ReturnValue = void> = (
-  edgeSpec: EdgeSpec,
-  port?: number
-) => ReturnValue
+// make this deeply immutable to force usage through helper functions
+export type EdgeSpecRouteBundle = ReadonlyDeep<EdgeSpecOptions>
```

Comment:
when does this need to be mutable? i.e. can we remove the export above? or if not maybe renaming the above type to `MutableEdgeSpecRouteBundle` would help

Reply:
from what I see maybe `EdgeSpecOptions` can be simplified to 
```ts
export interface EdgeSpecOptions {
  handleModuleServiceRouteNotFound
  handle404
}
```
?

Reply:
👍